### PR TITLE
add epoch to sequence cache keys

### PR DESCRIPTION
**Pull request by tbg**
this allows for poisoning of response cache entries as
outlined in RFC #3100 (txn gc).

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3144)

### CODE REVIEW

**Comment by tamird** [comment]
Reviewed 6 of 6 files at r1, 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[storage/replica.go, line 1039 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3144#-K3HCP52Usv4DwT55KxV)** ([raw file](https://github.com/cockroachdb/cockroach/blob/52b42cb1f9df056562744ae1956c76fa7c2a0ece/storage/replica.go#L1039)):</sup>
why do you need this edge case? isn't `sequence == 0` equivalent to `sequence < ba.Txn.Sequence`?


---

<sup>**[storage/replica_test.go, line 1557 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3144#-K3HCnRqTQr_S8IalEf_)** ([raw file](https://github.com/cockroachdb/cockroach/blob/52b42cb1f9df056562744ae1956c76fa7c2a0ece/storage/replica_test.go#L1557)):</sup>
seems like you'd normally introduce a scope here to avoid overwriting `err` a few lines down.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3144)

File: storage/replica_test.go [hunk 1]
```diff
@@ -1550,13 +1550,27 @@ func TestRangeSequenceCacheStoredTxnRetryError(t *testing.T) {
...
-	secondErr := try()
-	if _, ok := secondErr.(*roachpb.TransactionRetryError); !ok {
-		t.Fatal(secondErr)
+	err := try()
+	if _, ok := err.(*roachpb.TransactionRetryError); !ok {
+		t.Fatal(err)
+	}
+
+	//  Pretend we restarted by increasing the epoch. We didn't increase
+	//  the sequence though, so still the same error.
```
Comment (bdarnell) [inline storage/replica_test.go]: Why does that still get the same error? Aren't sequence numbers only valid during a single epoch?

File: storage/sequence_cache_test.go [hunk 2]
```diff
@@ -47,6 +47,8 @@ func createTestSequenceCache(t *testing.T, rangeID roachpb.RangeID, stopper *sto
 	return NewSequenceCache(rangeID), engine.NewInMem(roachpb.Attributes{}, 1<<20, stopper)
 }
 
+const testTxnEpo = 5
```
Comment (bdarnell) [inline storage/sequence_cache_test.go]: `s/Epo\b/Epoch/g`

**Comment by bdarnell** [comment]
LGTM

**Comment by tschottdorf** [comment]
Review status: 9 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[storage/replica.go, line 1039 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3144#-K3HCP52Usv4DwT55KxV)** ([raw file](https://github.com/cockroachdb/cockroach/blob/52b42cb1f9df056562744ae1956c76fa7c2a0ece/storage/replica.go#L1039)):</sup>
yeah, this is a remnant. Removed.


---

<sup>**[storage/replica_test.go, line 1563 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3144#-K3H8Zozv6NyWf-WZRx3-r2-1563)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d302b8b5512527121391ce928254acdb0f875391/storage/replica_test.go#L1563)):</sup>
It's easier to just have sequence numbers increasing throughout the lifetime of the transaction. The case in which the epoch increases but the sequence does not does not occur in practice, but it seemed worthwhile documenting it in a test.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3144)

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3144)

**Pull request merged**